src/main/java/org/opentripplanner/transit/raptor/heuristic/DefaultHeuristicRoutingStrategy.java

```diff
@@ -0,0 +1,309 @@
+      currentBoardingTotalDuration = previousDuration + slack;
+
+      // TODO: check what types of constrained transfers are available
+      final int boardCost = hasConstrainedTransfer
+        ? 0
+        : costCalculator.boardingCost(
+          roundProvider.round() < 2,
+          previousDuration,
+          stopIndex,
+          previousDuration + slack,
+          currentTrip,
+          // TODO add best constraint to heuristic trip
+          RaptorTransferConstraint.REGULAR_TRANSFER
+        );
+
+      //  TODO, we should add cost + time to an array, so we can check both what is the minimum time
+      //  and cost separately at arrival
+      currentBoardCost = bestCostPreviousRound(stopIndex) + boardCost;
+    }
+  }
+
+  @Override
+  public void alight(int stopIndex, int stopPos, int alightSlack) {
+    if (currentBoardingTime != UNREACHED) {
+      final int transitDuration = currentBoardingTime - currentTrip.departure(stopPos);
+      final int transitCost = costCalculator.transitArrivalCost(
+        currentBoardCost,
+        alightSlack,
+        transitDuration,
+        currentTrip,
+        stopIndex
+      );
+
+      updateNewBestTimeCostAndRound(
+        stopIndex,
+        currentBoardingTotalDuration + alightSlack + transitDuration,
+        currentBoardCost + transitCost,
+        true
+      );
+    }
+  }
+
+  @Override
+  public void transferToStops(int fromStop, Iterator<? extends RaptorTransfer> transfers) {
+    final int prevDuration = bestTimePreviousRound(fromStop);
+    final int prevCost = bestCostPreviousRound(fromStop);
+
+    while (transfers.hasNext()) {
+      RaptorTransfer it = transfers.next();
+      updateNewBestTimeCostAndRound(
+        it.stop(),
+        it.durationInSeconds() + prevDuration,
+        it.generalizedCost() + prevCost,
+        false
+      );
+    }
+  }
+
+  @Override
+  public void setAccessToStop(RaptorAccessEgress accessPath) {
+    updateNewBestTimeCostAndRound(
+      accessPath.stop(),
+      accessPath.durationInSeconds(),
+      accessPath.generalizedCost(),
+      accessPath.stopReachedOnBoard()
+    );
+  }
+
+  public Heuristics heuristics() {
+    return new ReverseHeuristicsAdapter(bestNumOfTransfers, times, costs, egressStops);
+  }
+
+  /**
+   * Clear all reached flags before we start a new iteration. This is important so stops visited in
+   * the previous iteration in the last round does not "overflow" into the next iteration.
+   */
+  private void setupIteration() {
+    // clear all touched stops to avoid constant reëxploration
+    reachedCurrentRound.clear();
+    reachedByTransitCurrentRound.clear();
+  }
+
+  private void swapReachedCurrentAndLastRound() {
+    BitSet tmp = reachedLastRound;
+    reachedLastRound = reachedCurrentRound;
+    reachedCurrentRound = tmp;
+  }
+
+  private void updateNewBestTimeCostAndRound(int stop, int time, int cost, boolean isTransit) {
+    if (heuristicRounds[stop] + roundProvider.round() > roundMaxLimit) {
```

Comment:
this check could be potentially done before alighting/transfering to avoid computing time and cost